File: ios/Nabi/AppProtocol.m
#import "AppProtocol.h"

@implementation AppProtocol

+ (BOOL)canInitWithRequest:(NSURLRequest *)request {
  return [request.URL.scheme isEqualToString:@"app"];
}

+ (NSURLRequest *)canonicalRequestForRequest:(NSURLRequest *)request {
  return request;
}

- (instancetype)initWithRequest:(NSURLRequest *)request cachedResponse:(NSCachedURLResponse *)cachedResponse client:(id<NSURLProtocolClient>)client {
  self = [super initWithRequest:request cachedResponse:cachedResponse client:client];
  return self;
}

- (void)startLoading {
  NSURL *url = self.request.URL;
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    NSString *fileName = [NSString stringWithFormat:@"%@/web/%@%@", [[NSBundle mainBundle] bundlePath], url.host, url.path];
    NSLog(@"loading %@ --> %@", url.absoluteString, fileName);
    NSData *data = [NSData dataWithContentsOfFile:fileName];
    NSString *ext = fileName.pathExtension;

    NSURLResponse *response = [[NSURLResponse alloc] initWithURL:self.request.URL
                                                        MIMEType:[self determineMIMETypeByExtension:ext]
                                           expectedContentLength:data.length
                                                textEncodingName:[self isExtensionTextFile:ext] ? @"utf-8" : nil];
    
    dispatch_async(dispatch_get_main_queue(), ^{
      [self.client URLProtocol:self didReceiveResponse:response cacheStoragePolicy:NSURLCacheStorageNotAllowed];
      [self.client URLProtocol:self didLoadData:data];
      [self.client URLProtocolDidFinishLoading:self];
    });
  });
}

- (void)stopLoading {
}

- (NSString *)determineMIMETypeByExtension:(NSString *)ext {
  if ([ext isEqualToString:@"css"]) {
    return @"text/css";
  } else if ([ext isEqualToString:@"gif"]) {
    return @"image/gif";
  } else if ([ext isEqualToString:@"html"] || [ext isEqualToString:@"htm"]) {
    return @"text/html";
  } else if ([ext isEqualToString:@"jpeg"] || [ext isEqualToString:@"jpg"]) {
    return @"image/jpeg";
  } else if ([ext isEqualToString:@"js"]) {
    return @"application/javascript";
  } else if ([ext isEqualToString:@"json"] || [ext isEqualToString:@"map"]) {
    return @"application/json";
  } else if ([ext isEqualToString:@"png"]) {
    return @"image/png";
  } else if ([ext isEqualToString:@"svg"]) {
    return @"image/svg+xml";
  } else if ([ext isEqualToString:@"text"] || [ext isEqualToString:@"txt"]) {
    return @"text/plain";
  } else if ([ext isEqualToString:@"ttf"]) {
    return @"application/font-sfnt";
  } else if ([ext isEqualToString:@"woff"]) {
    return @"application/font-woff";
  }
  return @"application/octet-stream";
}

- (BOOL)isExtensionTextFile:(NSString *)ext {
  NSArray *textExts = @[@"css", @"html", @"html", @"js", @"json", @"map", @"svg", @"text", @"txt"];
  return [textExts indexOfObject:ext] != NSNotFound;
}

@end

File: ios/Nabi/AppProtocol.swift
import UIKit

class AppProtocol: URLProtocol {
  override class func canInit(with request: URLRequest) -> Bool {
    return request.url!.scheme == "app"
  }

  override class func canonicalRequest(for request: URLRequest) -> URLRequest {
    return request;
  }

  override func startLoading() {
    let url = self.request.url!
    DispatchQueue.global(qos: .userInitiated).async {
      let fileName = "\(Bundle.main.bundlePath)/web/\(url.host!)\(url.path)"
      print("loading \(url.absoluteString) --> \(fileName)")
      let fileURL = URL(fileURLWithPath: fileName)

      let data = try? Data(contentsOf: fileURL)
      let ext = fileURL.pathExtension

      let response = URLResponse.init(url: self.request.url!,
                                      mimeType: mimeType(ext: ext),
                                      expectedContentLength: data!.count,
                                      textEncodingName: isTextFile(ext: ext) ? "utf-8" : nil)

      DispatchQueue.main.async {
        let client = self.client!
        client.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
        client.urlProtocol(self, didLoad: data!)
        client.urlProtocolDidFinishLoading(self)
      }
    }
  }

  override func stopLoading() {
  }
}

func mimeType(ext: String) -> String {
  switch ext {
  case "css":
    return "text/css"
  case "gif":
    return "image/gif"
  case "html", "htm":
    return "text/html"
  case "jpeg", "jpg":
    return "image/jpeg"
  case "js":
    return "application/javascript"
  case "json", "map":
    return "application/json"
  case "png":
    return "image/png"
  case "svg":
    return "image/svg+xml"
  case "text", "txt":
    return "text/plain"
  case "ttf":
    return "application/font-sfnt"
  case "woff":
    return "application/font-woff"
  default:
    return "application/octet-stream"
  }
}

func isTextFile(ext: String) -> Bool {
  return ["css", "html", "htm", "js", "json", "map", "svg", "text", "txt"].contains(ext)
}
